CMake improvements for consuming it via `add_subdirectory()`

vadz opened the pull request:
This started as [this thread](https://groups.google.com/g/wx-dev/c/XvuI422sGE8) but grew out of it to contain several QoL improvements for consuming the library from a superproject using CMake.

cc @MaartenBent

vadz committed:
Add wx_option_auto() CMake helper function

This just calls wx_option() to create an option taking ON/OFF/AUTO
values.

vadz committed:
Be more helpful if wxBUILD_LOCALES=ON and gettext not found

Explain that in addition to installing gettext the error can also be
avoided by not explicitly requesting translations to be built.

vadz committed:
Stop configuring if wxUSE_SECRETSTORE=ON but libsecret not found

This is similar to dd745f1e37 (Error in configure if wxSecretStore
requested but not available, 2025-05-27) but for CMake: allow requesting
wxSecretStore to be available by setting wxUSE_SECRETSTORE=ON explicitly
and failing in this case if libsecret is not found.

vadz committed:
Fix typo in wxUSE_REPRODUCIBLE_BUILD option description

vadz committed:
Define HAVE_UNISTD_H when building bundled zlib with CMake

Otherwise <unistd.h> is not included, resulting in many gcc
-Wimplicit-function-declaration warnings, so include wx/setup.h, where
configure will have defined this symbol if the header is available.

vadz committed:
Don't give summary at the end of CMake if nothing changed

This avoids annoyingly redisplaying the same summary again and again
when CMake is rerun due to some change to its files -- showing the
summary again in this case is not useful, especially when wxWidgets is
consumed by another project using CMake via add_subdirectory() and the
change happens in some CMake file not related to wxWidgets at all.

vadz committed:
Use new wx_lib_include_directories_before() it for 3rd party libs

Change wx_lib_include_directories() CMake helper to append include
directories to the path and add wx_lib_include_directories_before()
to prepend them to it in order to ensure that built-in 3rd party
libraries directories come in front of the headers search path.

This ensures that when src/common/imagpng.cpp includes png.h, the
correct src/png/pnglibconf.h is used and not any pnglibconf.h from the
system package, which wouldn't define PNG_PREFIX as "wx_", for example,
resulting in link errors later.

MaartenBent commented on build/cmake/functions.cmake:
```diff
@@ -706,6 +706,17 @@ function(wx_add_thirdparty_library var_name lib_name help_str)
+  # from the cached value.
+  foreach(entry IN LISTS wxTHIRD_PARTY_LIBRARIES)
+    # TODO-CMAKE-VERSION: Use list(JOIN) with CMake 3.12+.
+    string(APPEND wxTHIRD_PARTY_SUMMARY_NOW "${entry}-")
```
This doesn't work. `wxTHIRD_PARTY_LIBRARIES` is a list where the uneven entries are the variable name and the even entries the description, like:
```
0 - wxUSE_REGEX
1 - enable support for wxRegEx class
2 - wxUSE_ZLIB
3 - use zlib for LZW compression
4 - wxUSE_EXPAT
5 - use expat for XML parsing
...
```

I think you want the cached string to look like:
 ```
wxUSE_REGEX=builtin wxUSE_ZLIB=builtin wxUSE_EXPAT=builtin wxUSE_LIBJPEG=builtin wxUSE_LIBPNG=builtin wxUSE_LIBTIFF=builtin wxUSE_LIBWEBP=builtin wxUSE_NANOSVG=builtin wxUSE_LIBLZMA=OFF wxUSE_WEBVIEW=ON 
```

Then you'll need to do something similar as is done later in this function, like this:
```cmake
  set(var_name)
  foreach(entry IN LISTS wxTHIRD_PARTY_LIBRARIES)
    if(NOT var_name)
      set(var_name ${entry})
    else()
      string(APPEND wxTHIRD_PARTY_SUMMARY_NOW "${var_name}=${${var_name}} ")
      set(var_name)
    endif()
  endforeach()
```

But now the function iterates `wxTHIRD_PARTY_LIBRARIES` twice. Maybe it is simpler to cache the value of `${message}` and compare against that. At the expense of always creating the full print.

vadz replied:
Oops, thanks, I've totally missed this.

I've changed the code to build the summary cache as part of the first iteration (the code already iterated twice, I had added a third one), please let me know if you still see any problems with it.

MaartenBent commented on build/cmake/functions.cmake:
```diff
@@ -706,6 +706,17 @@ function(wx_add_thirdparty_library var_name lib_name help_str)
+    string(APPEND wxTHIRD_PARTY_SUMMARY_NOW "${entry}-")
+  endforeach()
+
+  if(NOT "${wxTHIRD_PARTY_SUMMARY_NOW}" STREQUAL "$CACHE{wxTHIRD_PARTY_SUMMARY}")
```
`$CACHE{VAR}` is added in version 3.13. I don't think it is needed here and you can just use `"${wxTHIRD_PARTY_SUMMARY}"`

vadz replied:
Thanks, I didn't notice this and I don't know why did I use `$CACHE{}` originally... I think I wanted to make sure I was using the original value of the variable, but it looks like it's not really necessary, so I've just removed it, thanks.

MaartenBent commented on build/cmake/main.cmake:
```diff
@@ -73,6 +73,20 @@ endif()
+    "${wxBUILD_TOOLKIT}-${wxTOOLKIT_VERSION}-${wxTOOLKIT_EXTRA}-"
+    "${wxBUILD_MONOLITHIC}-${wxBUILD_SHARED}-${wxBUILD_COMPATIBILITY}-"
+)
+if("${wxSUMMARY_NOW}" STREQUAL "$CACHE{wxSUMMARY}")
```
See `$CACHE{wxTHIRD_PARTY_SUMMARY}` comment

vadz replied:
Removed `CACHE` too.

MaartenBent commented on build/cmake/functions.cmake:
```diff
@@ -1025,6 +1048,15 @@ function(wx_option name desc)
+# just a warning otherwise, while ON means that an error is given if it can't
+# be enabled.
+function(wx_option_auto name desc)
+  wx_option(${name} ${desc} AUTO STRINGS ON OFF AUTO)
```
Our CMake files are mostly indented with 4 spaces.

MaartenBent commented on build/cmake/locale/CMakeLists.txt:
```diff
@@ -1,16 +1,20 @@
-elseif(wxBUILD_LOCALES)
-    find_package(Gettext REQUIRED)
+if(wxBUILD_LOCALES)
+    find_package(Gettext)
```
Why did you remove `QUIET` in the `AUTO` option? Now it pollutes the output every reconfigure with
```
Could NOT find Gettext (missing: GETTEXT_MSGMERGE_EXECUTABLE GETTEXT_MSGFMT_EXECUTABLE)
```

vadz replied:
Sorry, I wanted to simplify the code but didn't realize that it would start giving warnings in AUTO case when it's not available (I've only tested this here, where it is).

I've restored it now, I don't think using a temp variable and passing it to `find_package()` is better than doing it with just an `if()`, is it?

MaartenBent reviewed (changes requested):
> why is it a macro and not a function, BTW?

Functions are scoped, macros are not. With a function, `wxMONO_INCLUDE_DIRS` would be reset every time.

> Please let me know if you think this could/should be improved in some way.

I'm fine with this, but you could add a `BEFORE` parameter and use `cmake_parse_arguments`:

```cmake
wx_lib_include_directories(wxcore BEFORE ${${lib}_INCLUDE_DIRS})
```
```cmake
macro(wx_lib_include_directories name)
    cmake_parse_arguments(_INC_DIR "BEFORE" "" "" ${ARGN})
    if(wxBUILD_MONOLITHIC)
        list(APPEND wxMONO_INCLUDE_DIRS ${ARGN})
        set(wxMONO_INCLUDE_DIRS ${wxMONO_INCLUDE_DIRS} PARENT_SCOPE)
    else()
        if(_INC_DIR_BEFORE)
            target_include_directories(${name} BEFORE PRIVATE ${ARGN})
        else()
            target_include_directories(${name} PRIVATE ${ARGN})
        endif()
    endif()
endmacro()
```

vadz commented:
> I'm fine with this, but you could add a `BEFORE` parameter and use `cmake_parse_arguments`:

I _think_ I've tried something very similar to this and ran into problems, but I could have made some stupid mistake. Anyhow, if you're fine with the current version, let's leave it, but if you'd like to change it, please let me know or just do it as you prefer.

I've addressed your other comments in the new commits, thanks for the review!

vadz committed:
Use MSVC-specific define for zlib only when MSVC is being used

Test for the compiler, not platform, before defining
_CRT_NONSTDC_NO_WARNINGS: even if it does no harm when another compiler
is used, it's unnecessary.

MaartenBent reviewed (approved):
Looks good now!
